### PR TITLE
Expr to Sql : Case

### PR DESCRIPTION
**Pull request by yyy1000**
## Which issue does this PR close?

Related https://github.com/apache/arrow-datafusion/issues/9726.

### CODE REVIEW

File: datafusion/sql/src/unparser/expr.rs [hunk 1]
```diff
@@ -113,10 +113,38 @@ impl Unparser<'_> {
...
             }) => {
-                not_impl_err!("Unsupported expression: {expr:?}")
+                let conditions = when_then_expr
+                    .iter()
+                    .map(|(w, _)| self.expr_to_sql(w))
+                    .collect::<Result<Vec<_>>>()?;
+                let results = when_then_expr
+                    .iter()
+                    .map(|(_, t)| self.expr_to_sql(t))
+                    .collect::<Result<Vec<_>>>()?;
```
Comment (yyy1000) [inline datafusion/sql/src/unparser/expr.rs:126]: I'm trying to make these two iterations to one, but I failed. ☹️
I tried
1. `unzip`, but I got two vec which consist of `Result<Expr>` but not `Expr`
2. `match`, got type incompatible error 🥲
3. `collect`, but don't know how to collect a `(Result<Expr>, Result<Expr>)`

File: datafusion/sql/src/unparser/expr.rs [hunk 2]
```diff
@@ -698,17 +733,17 @@ mod tests {
                 }),
                 "COUNT(DISTINCT *)",
             ),
-            (Expr::IsNotNull(Box::new(col("a"))), r#""a" IS NOT NULL"#),
+            (col("a").is_not_null(), r#""a" IS NOT NULL"#),
```
Comment (yyy1000) [inline datafusion/sql/src/unparser/expr.rs:736]: I also changed some of these tests to make them looks concise.

File: datafusion/sql/src/unparser/expr.rs (hunk 1 above)
Comment (alamb) [reply datafusion/sql/src/unparser/expr.rs:126]: I likewise tried to use `izip` and came up with this, which wasn't particularly nice

```rust
                let (conditions, results): (Vec<_>, Vec<_>) = when_then_expr
                    .iter()
                    .map(|(w, t)| (self.expr_to_sql(w), self.expr_to_sql(t)))
                    .unzip();

                let conditions = conditions.into_iter().collect::<Result<Vec<_>>>()?;
                let results = results.into_iter().collect::<Result<Vec<_>>>()?;
```

This is the best I could come up with but I am not sure how much better it is:

```rust
                let mut conditions = vec![];
                let mut results = vec[]!;
                for (w, t) in when_then_expr {
                    conditions.push(self.expr_to_sql(w)?);
                    results.push(self.expr_to_sql(t)?);
                }
```

**Review by alamb — APPROVED** [review]
Looks good to me @yyy1000  -- thank you for the contribution.

File: datafusion/sql/src/unparser/expr.rs (hunk 1 above)
Comment (yyy1000) [reply datafusion/sql/src/unparser/expr.rs:126]: Yeah, I also tried two pushes before but I removed it for declaring two mut vec 😆

**Pull request merged**